Add sidekiq agent

bmarini opened the pull request:
This adds a metric for the length of each sidekiq queue. This also supports collecting stats for queues when using reliable fetch, where each queue has a sub queues per worker.

The following stats are collected:

```
sidekiq.processed
sidekiq.failed
sidekiq.scheduled_size
sidekiq.retry_size
sidekiq.dead_size
sidekiq.processes_size
sidekiq.workers_size

Per queue:
  sidekiq.queue.length tag=queue:<queue name>
  sidekiq.queue.latency tag=queue:<queue name>
```

poblahblahblah commented on checks.d/sidekiq.py:
```diff
@@ -0,0 +1,51 @@
+        self.gauge('sidekiq.redis.info.latency_ms', latency_ms, tags=tags)
+
+        # Check all sidekiq queue lengths
+        for queue in conn.smembers('queues'):
```
If you hardcode this as `queues` it doesn't allow for namespaced queues.

poblahblahblah commented:
I deployed this plugin to a single instance to test it out and left it running overnight and it looks like it strongly impacted redis performance and the datadog agent run time. Attached is a NR snapshot of our app before the plugin was installed, during the time the plugin was running, and after the plugin was removed. The big spike occurs when we deployed the single check, and the drop off occurs this morning when I removed the check.

redis version: 2.8.19
server os: ubuntu 12.04 x86_64
sidekiq version: 3.5.0
datadog agent version: 5.5.0
agent run time before sidekiq check: ~5.10 seconds
agent run time w/sidekiq check: ~16 seconds
average time spent in redis before sidekiq check: ~4ms
average time spent in redis w/sidekiq check: ~410ms

![screen shot 2015-11-11 at 8 45 44 am](https://cloud.githubusercontent.com/assets/51938/11096779/a3ec9844-8850-11e5-90fe-7c4d11e308a5.png)

bmarini commented:
Thanks for testing @poblahblahblah I'll have to dig in to see what the issue is. What is the easiest way for me to build the agent with my plugin added? I looked at building from source but the scripts seems to require a tag from the official repo...

bmarini commented:
Hey @poblahblahblah can you give it another try? I updated it to make less redis calls.

poblahblahblah commented:
@bmarini will do

poblahblahblah commented:
Hey @bmarini,

This checks looks better, but it still has a noticeable impact on redis response time:

![screen shot 2015-11-12 at 3 00 32 pm](https://cloud.githubusercontent.com/assets/51938/11133973/322801b8-894e-11e5-9405-29bfa4acafc1.png)

It looks like a big step in the right direction, though.

bmarini commented:
Hey @poblahblahblah if you don't mind testing it out another time. I moved everything into a lua script.

poblahblahblah commented:
Hey @bmarini, here are a few things I noticed:

`local scheduled_size = redis.call("GET", KEYS[3])`

should be

`local scheduled_size = redis.call("ZCARD", KEYS[3])`

as that key should be a `zset`

There was also an issue with this block:

```
        local workers_size = 0
        for i, pkey in ipairs(processes) do
          workers_size = workers_size + tonumber(redis.call("HGET", pkey, "busy"))
        end
```

the problem is that the process will have the namespace prepended to it, at least in our setup.

poblahblahblah commented:
@bmarini thanks for updating things, although unfortunately it looks like we are still seeing a bump in redis latency in New Relic after adding the check:

<img width="728" alt="screen shot 2015-11-16 at 5 43 01 pm" src="https://cloud.githubusercontent.com/assets/51938/11200265/82a3ee16-8c89-11e5-9f0e-56e7643c7c7e.png">


It would be great if someone else could give this a shot to verify that they are also seeing an increase.

clod81 commented:
:+1:  
any chance to have this available anytime soon in datadog integrations?

Poogles commented:
@bmarini You're using keys which is going to have a pretty big performance hit (check the warning on http://redis.io/commands/keys).  Have you tried using scan with match instead?  http://redis.io/commands/scan

irabinovitch commented:
@clod81 At the moment the integration is being reviewed and it looks like there maybe a few places where there is room for improvement based on @poblahblahblah 's comments. However, if you would like to test it out you should be able to install it locally as an agent check by downloading it and placing it in the checks.d folder.  We'd love to have your feedback on the check as well.

voxxit commented on checks.d/sidekiq.py:
```diff
@@ -0,0 +1,261 @@
+        local retry_size = redis.call("GET", KEYS[4])
+        local dead_size = redis.call("GET", KEYS[5])
+        local processes_size = redis.call("SCARD", KEYS[6])
+        local processes = redis.call("SMEMBERS", KEYS[6])
```
These could possibly be switched to `MGET` to save roundtrips.

voxxit commented on checks.d/sidekiq.py:
```diff
@@ -0,0 +1,261 @@
+    #
+    def _register_script(self):
+        s = """
+        local processed = redis.call("GET", KEYS[1])
```
These `GET` calls could possibly be switched to `MGET` calls in order to save roundtrips.

remh replied:
@bmarini what do you think ?

clod81 commented:
I'm getting an error when trying to run the agent check

My configuration is the following:

<pre>
init_config:

instances:
  - host: localhost 
    port: 6379
    namespace: __sdkq__
</pre>

<pre>
2015-11-25 14:36:39,002 | INFO | dd.collector | checks.collector(collector.py:529) | Running check sidekiq
2015-11-25 14:36:39,003 | ERROR | dd.collector | checks.sidekiq(__init__.py:762) | Check 'sidekiq' instance #0 failed
Traceback (most recent call last):
  File "/opt/datadog-agent/agent/checks/__init__.py", line 745, in run
    self.check(copy.deepcopy(instance))
  File "/etc/dd-agent/checks.d/sidekiq.py", line 261, in check
    self._check_sidekiq(instance, custom_tags)
  File "/etc/dd-agent/checks.d/sidekiq.py", line 239, in _check_sidekiq
    stats = Stats(conn, instance.get('namespace'))
  File "/etc/dd-agent/checks.d/sidekiq.py", line 18, in __init__
    self._check_stats()
  File "/etc/dd-agent/checks.d/sidekiq.py", line 126, in _check_stats
    res = self._run_script()
  File "/etc/dd-agent/checks.d/sidekiq.py", line 123, in _run_script
    self.key('queue:*')])
  File "/opt/datadog-agent/embedded/lib/python2.7/site-packages/redis/client.py", line 2651, in __call__
    return client.evalsha(self.sha, len(keys), *args)
  File "/opt/datadog-agent/embedded/lib/python2.7/site-packages/redis/client.py", line 1911, in evalsha
    return self.execute_command('EVALSHA', sha, numkeys, *keys_and_args)
  File "/opt/datadog-agent/embedded/lib/python2.7/site-packages/redis/client.py", line 565, in execute_command
    return self.parse_response(connection, command_name, **options)
  File "/opt/datadog-agent/embedded/lib/python2.7/site-packages/redis/client.py", line 577, in parse_response
    response = connection.read_response()
  File "/opt/datadog-agent/embedded/lib/python2.7/site-packages/redis/connection.py", line 574, in read_response
    raise response
ResponseError: Error running script (call to f_96500133b7d419997493f781af65f54356edc31d): @user_script:12: user_script:12: attempt to perform arithmetic on a nil value 
Metrics: 
[('sidekiq.redis.info.latency_ms',
  1448415399,
  0.46,
  {'hostname': 'i-9abbe65a',
   'tags': ['redis_host:localhost', 'redis_port:6379'],
   'type': 'gauge'})]
Events: 
[]
Service Checks: 
[{'check': 'sidekiq.redis.can_connect',
  'host_name': 'i-9abbe65a',
  'id': 1,
  'message': None,
  'status': 0,
  'tags': ['redis_host:localhost', 'redis_port:6379'],
  'timestamp': 1448415399.0035}]
Service Metadata: 
[{'version': '2.8.4'}]
    sidekiq
    -------
      - instance #0 [ERROR]: 'Error running script (call to f_96500133b7d419997493f781af65f54356edc31d): @user_script:12: user_script:12: attempt to perform arithmetic on a nil value '
        Traceback (most recent call last):
          File "/opt/datadog-agent/agent/checks/__init__.py", line 745, in run
            self.check(copy.deepcopy(instance))
          File "/etc/dd-agent/checks.d/sidekiq.py", line 261, in check
            self._check_sidekiq(instance, custom_tags)
          File "/etc/dd-agent/checks.d/sidekiq.py", line 239, in _check_sidekiq
            stats = Stats(conn, instance.get('namespace'))
          File "/etc/dd-agent/checks.d/sidekiq.py", line 18, in __init__
            self._check_stats()
          File "/etc/dd-agent/checks.d/sidekiq.py", line 126, in _check_stats
            res = self._run_script()
          File "/etc/dd-agent/checks.d/sidekiq.py", line 123, in _run_script
            self.key('queue:*')])
          File "/opt/datadog-agent/embedded/lib/python2.7/site-packages/redis/client.py", line 2651, in __call__
            return client.evalsha(self.sha, len(keys), *args)
          File "/opt/datadog-agent/embedded/lib/python2.7/site-packages/redis/client.py", line 1911, in evalsha
            return self.execute_command('EVALSHA', sha, numkeys, *keys_and_args)
          File "/opt/datadog-agent/embedded/lib/python2.7/site-packages/redis/client.py", line 565, in execute_command
            return self.parse_response(connection, command_name, **options)
          File "/opt/datadog-agent/embedded/lib/python2.7/site-packages/redis/client.py", line 577, in parse_response
            response = connection.read_response()
          File "/opt/datadog-agent/embedded/lib/python2.7/site-packages/redis/connection.py", line 574, in read_response
            raise response
        ResponseError: Error running script (call to f_96500133b7d419997493f781af65f54356edc31d): @user_script:12: user_script:12: attempt to perform arithmetic on a nil value 
        
      - Collected 1 metric, 0 events & 1 service check
</pre>

clod81 commented:
it works if I remove everything apart from stat:processed and stat:failed

<pre>
'''
Sidekiq checks
'''
# stdlib
import time

# 3rd party
import redis

# project
from checks import AgentCheck

class Stats:
    def __init__(self, conn, namespace):
        self.conn = conn
        self.namespace = namespace
        self._script = self._register_script()
        self._check_stats()

    def _register_script(self):
        s = """
        local processed = redis.call("GET", KEYS[1])
        local failed = redis.call("GET", KEYS[2])
        local result = {processed,failed}
        return result
        """
        return self.conn.register_script(s)

    def _run_script(self):
        return self._script(keys=[
                            self.key('stat:processed'),
                            self.key('stat:failed')])

    def _check_stats(self):
        res = self._run_script()
        self.processed = int(res[0])
        self.failed = int(res[1])

    def key(self, name):
        return "%s:%s" % (self.namespace, name) if self.namespace else name


class Sidekiq(AgentCheck):
    SOURCE_TYPE_NAME = 'sidekiq'

    def __init__(self, name, init_config, agentConfig, instances=None):
        AgentCheck.__init__(self, name, init_config, agentConfig, instances)
        self.connections = {}

    def _key(self, instance, name):
        ns = instance.get('namespace')
        return "%s:%s" % (ns, name) if ns else name

    # Taken from redisdb.py, returns custom tags and redis tags for redis connection.
    def _get_tags(self, custom_tags, instance):
        tags = set(custom_tags or [])

        if 'unix_socket_path' in instance:
            tags_to_add = [
                "redis_host:%s" % instance.get("unix_socket_path"),
                "redis_port:unix_socket",
            ]
        else:
            tags_to_add = [
                "redis_host:%s" % instance.get('host'),
                "redis_port:%s" % instance.get('port')
            ]

        tags = sorted(tags.union(tags_to_add))

        return tags, tags_to_add

    # Taken from redisdb.py, generates a uniq key to cache redis connections.
    def _generate_instance_key(self, instance):
        if 'unix_socket_path' in instance:
            return (instance.get('unix_socket_path'), instance.get('db'))
        else:
            return (instance.get('host'), instance.get('port'), instance.get('db'))

    # Taken from redisdb.py, collects metadata from the redis server.
    def _collect_metadata(self, info):
        if info and 'redis_version' in info:
            self.service_metadata('version', info['redis_version'])

    # Taken from redisdb.py, retrieves a cached redis connection.
    def _get_conn(self, instance):
        key = self._generate_instance_key(instance)
        if key not in self.connections:
            try:

                # Only send useful parameters to the redis client constructor
                list_params = ['host', 'port', 'db', 'password', 'socket_timeout',
                               'connection_pool', 'charset', 'errors', 'unix_socket_path']

                # Set a default timeout (in seconds) if no timeout is specified in the instance config
                instance['socket_timeout'] = instance.get('socket_timeout', 5)

                connection_params = dict((k, instance[k]) for k in list_params if k in instance)

                self.connections[key] = redis.Redis(**connection_params)

            except TypeError:
                raise Exception("You need a redis library that supports authenticated connections. Try sudo easy_install redis.")

        return self.connections[key]

    def _check_sidekiq(self, instance, custom_tags):
        conn = self._get_conn(instance)

        tags, tags_to_add = self._get_tags(custom_tags, instance)

        # Ping the database for info, and track the latency.
        # Process the service check: the check passes if we can connect to Redis
        start = time.time()
        info = None
        try:
            info = conn.info()
            status = AgentCheck.OK
            self.service_check('sidekiq.redis.can_connect', status, tags=tags_to_add)
            self._collect_metadata(info)
        except ValueError, e:
            status = AgentCheck.CRITICAL
            self.service_check('sidekiq.redis.can_connect', status, tags=tags_to_add)
            raise
        except Exception, e:
            status = AgentCheck.CRITICAL
            self.service_check('sidekiq.redis.can_connect', status, tags=tags_to_add)
            raise

        latency_ms = round((time.time() - start) * 1000, 2)
        self.gauge('sidekiq.redis.info.latency_ms', latency_ms, tags=tags)

        stats = Stats(conn, instance.get('namespace'))

        # Check global sidekiq stats
        self.gauge('sidekiq.processed', stats.processed, tags=tags)
        self.gauge('sidekiq.failed', stats.failed, tags=tags)

    def check(self, instance):
        if ("host" not in instance or "port" not in instance) and "unix_socket_path" not in instance:
            raise Exception("You must specify a host/port couple or a unix_socket_path")
        custom_tags = instance.get('tags', [])

        self._check_sidekiq(instance, custom_tags)
</pre>

Poogles commented:
@poblahblahblah Can you confirm how many keys you have in Redis?

https://github.com/remind101/dd-agent/commit/d8b00fd472abd8b188e165df95dc2104aefccb0f#diff-d18ebe09b53bdba6188b25fed20c5eddR70

Is going to block Redis until it returns, going off the docs and the increase in response time I'm guessing you've got somewhere around a million keys?  Just trying to confirm it's this before going any further.

poblahblahblah commented:
@Poogles `keys "queue:*"`, which I think is what you are asking me to check, returns an empty set, but it takes about 1.7 seconds to call it.

bmarini commented:
with standard sidekiq, `keys` isn't even necessary, as there is a set of all queues. But with sidekiq pro and reliable fetch, each worker creates a unique sub queue which is not stored in a set. I think the only way to make it work performantly with reliable fetch is to update sidekiq to track all sub queues in a set. Of course there are probably some complications there so I was hoping to avoid that. I'll update to support just vanilla sidekiq and circle back on sidekiq pro support, once I get back from vacation next week.

bmarini commented:
@poblahblahblah I updated based on your feedback, and removed the `keys` call. It no longer supports the reliable fetch feature of sidekiq-pro, but that can be fixed in a later PR.

poblahblahblah commented on checks.d/sidekiq.py:
```diff
@@ -0,0 +1,246 @@
+        local workers_size = 0
+        for i, pkey in ipairs(processes) do
+            nspkey = ARGV[1] == "None" and pkey or (ARGV[1] .. ":" .. pkey)
+            workers_size = workers_size + tonumber(redis.call("HGET", nspkey, "busy"))
```
Lua threw an error about `nspkey` and `workers_size` being global variables. Changing these to local variables fixed the error for me.

```
            local nspkey = ARGV[1] == "None" and pkey or (ARGV[1] .. ":" .. pkey)
            local workers_size = workers_size + tonumber(redis.call("HGET", nspkey, "busy"))
```

clod81 commented:
thanks, all working well with standard sidekiq now

irabinovitch commented:
Thanks all. Fantastic work.  Looking through it shortly.

remh commented on checks.d/sidekiq.py:
```diff
@@ -0,0 +1,246 @@
+# project
+from checks import AgentCheck
+
+class Stats:
```
Your class you inherit from Object.
http://stackoverflow.com/questions/7304056/why-inheriting-from-object-type

remh commented on checks.d/sidekiq.py:
```diff
@@ -0,0 +1,246 @@
+            status = AgentCheck.OK
+            self.service_check('sidekiq.redis.can_connect', status, tags=tags_to_add)
+            self._collect_metadata(info)
+        except ValueError, e:
```
looks like you are handling these two exceptions in the same way. As `except Exception` is a catch-all. You don't need the first catch

bmarini replied:
I copied this from the redis check, so might have to fix there as well

remh commented on checks.d/sidekiq.py:
```diff
@@ -0,0 +1,246 @@
+            raise
+
+        latency_ms = round((time.time() - start) * 1000, 2)
+        self.gauge('sidekiq.redis.info.latency_ms', latency_ms, tags=tags)
```
this metric would already be collected by the existing redis check. I don't think we should collect it here as well.

remh commented on checks.d/sidekiq.py:
```diff
@@ -0,0 +1,246 @@
+        self.gauge('sidekiq.workers_size', stats.workers_size, tags=tags)
+
+        # Check queue stats
+        for queue, stat in stats.queue_stats.iteritems():
```
Can you implement a hard limit and a whitelisting system ? A given integration shouldn't create an unbounded number of metrics.

irabinovitch replied:
@bmarini The Couch integration has an example of a whitelist that might be a good starting point.
https://github.com/DataDog/dd-agent/pull/2066#discussion_r47966153

Theres an example of a hard limit there too.

remh commented on checks.d/sidekiq.py:
```diff
@@ -0,0 +1,246 @@
+        AgentCheck.__init__(self, name, init_config, agentConfig, instances)
+        self.connections = {}
+
+    def _key(self, instance, name):
```
It seems that this method is not called anywhere

remh commented on checks.d/sidekiq.py:
```diff
@@ -0,0 +1,246 @@
+                self.connections[key] = redis.Redis(**connection_params)
+
+            except TypeError:
+                raise Exception("You need a redis library that supports authenticated connections. Try sudo easy_install redis.")
```
Shouldn't be needed, the Agent bundles a pinned version of the python redis library. Currently 2.10.3 https://github.com/DataDog/omnibus-software/blob/master/config/software/python-redis.rb

remh commented on checks.d/sidekiq.py:
```diff
@@ -0,0 +1,246 @@
+    def check(self, instance):
+        if ("host" not in instance or "port" not in instance) and "unix_socket_path" not in instance:
+            raise Exception("You must specify a host/port couple or a unix_socket_path")
+        custom_tags = instance.get('tags', [])
```
I think you should extract the custom tags in your `_get_tags` method. 
This way you won't have to pass the custom tags all along as you are already passing the whole instance dict.

remh commented on checks.d/sidekiq.py:
```diff
@@ -0,0 +1,246 @@
+
+        offset = 8
+        self.queue_stats = {}
+        for i in range(num_queues):
```
nitpick, you should use `xrange` instead of `range`

remh replied:
Also looks like you could use generator between here and when you actually send the metrics. This way you wouldn't have to actually store all of these data.

remh commented on tests/checks/integration/test_sidekiq.py:
```diff
@@ -0,0 +1,124 @@
+# stdlib
+import logging
+import time
+import json
```
nitpick but you could use simplejson

remh commented:
Thanks a lot @bmarini !
It's a great PR with proper tests and CI!

I made a few comments but this looks already nice. Can you take a pass at them please ?

irabinovitch commented:
@bmarini Have you by any chance had an opportunity to review @remh 's comments? I'd love to see this go into 5.7, but our merge window is closing shortly.

irabinovitch commented:
@bmarini Friendly ping. Just wanted to check back in and see if you have a moment to help address the comments from code review?

bmarini commented:
@irabinovitch Sorry, I moved teams at work and am no longer working on the datadog integration. We've been running this in prod successfully though, since I opened this PR. If someone else can take this over and push it through please do!

spinx commented:
We'd need this before looking deploying Datadog ... Is there a timeline for this to be released?

parabuzzle commented:
Is there an ETA for this?

parabuzzle commented:
It appears that the tests are all passing except the `FLAVOR=zookeeper` and it appears to be related to the framework and not these changes:

```
Failed task: RuntimeError Command failed with status (2): [tar zxf $VOLATILE_DIR/zookeeper-3.4.6.tar....]
```

irabinovitch commented:
@parabuzzle This is currently scheduled for inclusion in the 5.8 release. In the mean time we've heard great community feedback from users downloading and installing the check via the checks.d folder.

remh commented:
@parabuzzle There are a few comments that need to be addressed. As mentioned by @irabinovitch you can use this check as a custom check but for now it's not ready to be merged unfortunately.

bmandrade commented:
Hi guys,

I'm really interested in this check for datadog, but I would like to ask if it's possible to add multiple namespaces to monitor, instead of just one?

Cheers 👍

prashanthvenkat93 commented:
Hi guys,
I would really love to see this PR go through. can somebody please take this up ?

bronislav commented:
@bmandrade I think multiple namespaces is a very rare use case. As a workaround, you may configure several instances with same host/port pairs, but different namespaces

bronislav commented on checks.d/sidekiq.py:
```diff
@@ -0,0 +1,246 @@
+        local workers_size = 0
+        for i, pkey in ipairs(processes) do
+            local nspkey = ARGV[1] == "None" and pkey or (ARGV[1] .. ":" .. pkey)
+            workers_size = workers_size + tonumber(redis.call("HGET", nspkey, "busy"))
```
I've notice that script fails on this line time to time, with `user_script:13: attempt to perform arithmetic on a nil value`. This is because `tonumber` on `nil` returns `nil`. After changing this to `(tonumber(redis.call("HGET", nspkey, "busy")) or 0)` fails stopped.

truthbk commented:
Hi @bmarini , since the transition to the SDK has begun I have taken the liberty to open a PR with your work here: DataDog/integrations-extras#32. You authorship should be preserved.

This should hopefully ease the pain of transitioning to the SDK. Let me know if you have any questions.

bmarini commented:
Thanks @truthbk for that! Much appreciated.

irabinovitch commented:
@bmarini @poblahblahblah Would it be ok with you if we close this PR and continue discussions over at https://github.com/DataDog/integrations-extras/issues/32 ? We are eager to merge and release sidekiq support.

We are currently in the process of migrating agent checks from this repository into two new repositories  https://github.com/DataDog/integrations-core and https://github.com/DataDog/integrations-extras in order to support a new integration release process for integrations.  This change will allow us to more quickly review, test and merge new integrations such as this one here.  

I realize this PR has been in the works for quite some time and apologize that it has taken us so long to merge.  Now that we are completing this migration of repositories, we will be able to release new integrations at a much faster pace. 

We are committed to being responsive and supportive as you start work in the new repos.  To facilitate the migration we've created a few resources for you:
- [Integration SDK Guide](http://docs.datadoghq.com/guides/integration_sdk/) : A guide that documents the new development and release model for integrations to the dd-agent.
- [datadog-sdk-testing]( https://github.com/DataDog/datadog-sdk-testing) gem: A library to ease the testing process for integrations.
- [community office hours](https://github.com/DataDog/dd-agent/wiki/Community-Office-Hours) where you can stop by for assistance with your integration PRs. 

Thank you for your patience as we have been working through this transition.  If you have any questions, we are happy to discuss here in this PR or on [Slack](http://chat.datadoghq.com) in #integrations.

irabinovitch commented:
@bmarini @poblahblahblah  Just a quick note that we're holding our regularly scheduled [community office hours and code review](https://github.com/DataDog/dd-agent/wiki/Community-Office-Hours) if you want to join us and discuss this PR.

masci commented:
Let's continue the discussion here https://github.com/DataDog/integrations-extras/pull/32, closing this
